### PR TITLE
Forall Context

### PR DESCRIPTION
**Pull request by zhang677**
A new concept "Forall Context" is introduced to solve the "= or +=" choice in workspace. Test "tile_dotProduct_1" is enabled now!

### CODE REVIEW

**Comment by zhang677** [comment]
Update! A new schedule command `wsaccel` is added to control whether the workspace acceleration is used and to assign which dimensions the workspace acceleration is used. Test "tile_dotProduct_1" is enabled now!

File: src/index_notation/index_notation.cpp [hunk 1]
```diff
@@ -2476,6 +2546,8 @@ struct TensorVar::Content {
   Format format;
   Schedule schedule;
   Literal fill;
+  std::vector<IndexVar> accels;
```
Comment (weiya711) [inline src/index_notation/index_notation.cpp:2549]: Again, change variable names here

File: src/index_notation/index_notation.cpp [hunk 2]
```diff
@@ -2508,6 +2580,8 @@ TensorVar::TensorVar(const int& id, const string& name, const Type& type, const
   content->type = type;
   content->format = format;
   content->fill = fill.defined()? fill : Literal::zero(type.getDataType());
+  content->accels = std::vector<IndexVar> {};
```
Comment (weiya711) [inline src/index_notation/index_notation.cpp:2583]: Again, change variable names here

File: src/index_notation/index_notation.cpp [hunk 3]
```diff
@@ -2551,6 +2625,19 @@ const Literal& TensorVar::getFill() const {
   return content->fill;
 }
 
+const std::vector<IndexVar>& TensorVar::getAccels() const {
+  return content->accels;
```
Comment (weiya711) [inline src/index_notation/index_notation.cpp:2629]: Again, change variable names here

File: test/tests-scheduling-eval.cpp [hunk 4]
```diff
@@ -154,7 +154,7 @@ IndexStmt scheduleTTMCPU(IndexStmt stmt, Tensor<double> B, int CHUNK_SIZE=16, in
 }
 
 IndexStmt scheduleMTTKRPCPU(IndexStmt stmt, Tensor<double> B, int CHUNK_SIZE=16, int UNROLL_FACTOR=8) {
-  int NUM_J = 1039/20;
+  // int NUM_J = 1039/20;
```
Comment (weiya711) [inline test/tests-scheduling-eval.cpp]: I don't believe this should be commented out

File: test/tests-workspaces.cpp [hunk 5]
```diff
@@ -516,8 +516,8 @@ TEST(workspaces, DISABLED_tile_dotProduct_2) {
   Tensor<double> C("C", {N}, Format({Dense}));
 
   for (int i = 0; i < N; i++) {
-    B.insert({i}, (double) i);
-    C.insert({i}, (double) i);
+    B.insert({i}, (double) i / N);
```
Comment (weiya711) [inline test/tests-workspaces.cpp:519]: Why was this necessary?

**Review by weiya711 — CHANGES_REQUESTED** [review]
I have added review comments that should be fixed before merging

File: test/tests-workspaces.cpp (hunk 5 above)
Comment (zhang677) [reply test/tests-workspaces.cpp:519]: Because current TACO checking function can't correctly check the equality between two relatively large numbers. If B's content is not divided by N, the test will fail even if it is correct

File: include/taco/index_notation/index_notation.h [hunk 6]
```diff
@@ -1143,6 +1187,15 @@ class TensorVar : public util::Comparable<TensorVar> {
   /// Gets the fill value of the tensor variable. May be left undefined.
   const Literal& getFill() const;
 
+  /// Gets the acceleration dimensions
+  const std::vector<IndexVar>& getaccelIndexVars() const;
+
+  /// Gets the acceleration flag
+  bool getshouldAccel() const;
```
Comment (weiya711) [inline include/taco/index_notation/index_notation.h:1194]: `getShouldAccel` and `setAccelIndexVars` should be capitalized as shown

**Review by weiya711 — CHANGES_REQUESTED** [review]

**Review by weiya711 — APPROVED** [review]
